packages/contracts/contracts/SmartInvoiceFeeEscrow.sol

```diff
@@ -0,0 +1,561 @@
+            "resolution != remainder"
+        );
+
+        if (_providerAward > 0) {
+            _transferPayment(token, provider, _providerAward);
+        }
+        if (_clientAward > 0) {
+            IERC20(token).safeTransfer(client, _clientAward);
+        }
+        if (resolutionFee > 0) {
+            IERC20(token).safeTransfer(resolver, resolutionFee);
+        }
+
+        milestone = amounts.length;
+        locked = false;
+
+        emit Resolve(
+            _msgSender(),
+            _clientAward,
+            _providerAward,
+            resolutionFee,
+            _details
+        );
+    }
+
+    /**
+     * @dev External function to unlock the contract.
+     * @param _disputeId The ID of the dispute
+     * @param _ruling The ruling of the arbitrator
+     */
+    function rule(uint256 _disputeId, uint256 _ruling)
+        external
+        virtual
+        override
+        nonReentrant
+    {
+        // called by arbitrator
+        require(resolverType == ADR.ARBITRATOR, "!arbitrator resolver");
+        require(locked, "!locked");
+        require(_msgSender() == resolver, "!resolver");
+        require(_disputeId == disputeId, "incorrect disputeId");
+        require(_ruling <= NUM_RULING_OPTIONS, "invalid ruling");
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        require(balance > 0, "balance is 0");
+
+        uint8[2] memory ruling = _getRuling(_ruling);
+        uint8 clientShare = ruling[0];
+        uint8 providerShare = ruling[1];
+        uint8 denom = clientShare + providerShare;
+        uint256 providerAward = (balance * providerShare) / denom;
+        uint256 clientAward = balance - providerAward;
+
+        if (providerAward > 0) {
+            _transferPayment(token, provider, providerAward);
+        }
+        if (clientAward > 0) {
+            _transferPayment(token, client, clientAward);
+        }
+
+        milestone = amounts.length;
+        locked = false;
+
+        emit Rule(resolver, clientAward, providerAward, _ruling);
+        emit Ruling(resolver, _disputeId, _ruling);
+    }
+
+    /**
+     * @dev Internal function to get the ruling of the arbitrator.
+     * @param _ruling The ruling of the arbitrator
+     */
+    function _getRuling(uint256 _ruling)
+        internal
+        pure
+        returns (uint8[2] memory ruling)
+    {
+        uint8[2][6] memory rulings = [
+            [1, 1], // 0 = refused to arbitrate
+            [1, 0], // 1 = 100% to client
+            [3, 1], // 2 = 75% to client
+            [1, 1], // 3 = 50% to client
+            [1, 3], // 4 = 25% to client
+            [0, 1] // 5 = 0% to client
+        ];
+        ruling = rulings[_ruling];
+    }
+
+    function _calculateFee(uint256 _amount)
+        internal
+        view
+        returns (uint256 fee, uint256 adjustedAmount)
+    {
+        fee = (_amount * feePercentage) / 100;
+        adjustedAmount = _amount - fee;
+        return (fee, adjustedAmount);
+    }
```

Comment:
I feel like this should be part of the FeeManager rather than in the escrow implementation.

Reply:
Yes that makes sense, especially if there are more complicated fee structures down the road that are reused among different types

Reply:
The issue here is that we don't want to change the fee on an existing invoice, so we'll have to make things immutable.